src/ILLink.CodeFix/BaseAttributeCodeFixProvider.cs

```diff
@@ -32,55 +32,52 @@ public sealed override FixAllProvider GetFixAllProvider ()
-				var symbol = semanticModel!.Compilation.GetTypeByMetadataName (FullyQualifiedAttributeName);
-				var document = context.Document;
-				var editor = new SyntaxEditor (root, document.Project.Solution.Workspace);
-				var generator = editor.Generator;
-				var attrArgs = GetAttributeArguments (semanticModel, targetNode, declarationSyntax, generator, diagnostic);
-				var codeFixTitle = CodeFixTitle.ToString ();
-
-				// Register a code action that will invoke the fix.
-				context.RegisterCodeFix (
-					CodeAction.Create (
-						title: codeFixTitle,
-						createChangedDocument: c => AddAttribute (
-							document, editor, generator, declarationSyntax, attrArgs, symbol!, c),
-						equivalenceKey: codeFixTitle),
-					diagnostic);
-			}
+			SyntaxNode targetNode = root!.FindNode (diagnostic.Location.SourceSpan);
+			if (FindAttributableParent (targetNode, AttributableParentTargets) is not SyntaxNode attributableNode)
+				return;
+
+			var model = await document.GetSemanticModelAsync (context.CancellationToken).ConfigureAwait (false);
+			var targetSymbol = model!.GetSymbolInfo (targetNode).Symbol!;
+
+			var attributableSymbol = model!.GetDeclaredSymbol (attributableNode)!;
+			var attributeSymbol = model!.Compilation.GetTypeByMetadataName (FullyQualifiedAttributeName)!;
+			var attributeArguments = GetAttributeArguments (SyntaxGenerator.GetGenerator (document), attributableSymbol, targetSymbol, diagnostic);
+			var codeFixTitle = CodeFixTitle.ToString ();
+
+			context.RegisterCodeFix (CodeAction.Create (
+				title: codeFixTitle,
+				createChangedDocument: ct => AddAttributeAsync (
+					document, attributableNode, attributeArguments, attributeSymbol, ct),
+				equivalenceKey: codeFixTitle), diagnostic);
 		}
 
-		private static async Task<Document> AddAttribute (
+		private static async Task<Document> AddAttributeAsync (
 			Document document,
-			SyntaxEditor editor,
-			SyntaxGenerator generator,
-			CSharpSyntaxNode containingDecl,
-			SyntaxNode[] attrArgs,
-			ITypeSymbol AttributeSymbol,
+			SyntaxNode targetNode,
+			SyntaxNode[] attributeArguments,
+			ITypeSymbol attributeSymbol,
 			CancellationToken cancellationToken)
 		{
-			var semanticModel = await document.GetSemanticModelAsync (cancellationToken).ConfigureAwait (false);
-			if (semanticModel is null)
-				return document;
-
-			var newAttribute = generator
-				.Attribute (generator.TypeExpression (AttributeSymbol), attrArgs)
-				.WithAdditionalAnnotations (
-					Simplifier.Annotation,
-					Simplifier.AddImportsAnnotation);
+			var editor = await DocumentEditor.CreateAsync (document, cancellationToken).ConfigureAwait (false);
+			var generator = editor.Generator;
+			var attribute = CreateNewAttribute (generator, attributeSymbol, attributeArguments);
+			editor.AddAttribute (targetNode, attribute);
+			return editor.GetChangedDocument ();
+		}
 
-			editor.AddAttribute (containingDecl, newAttribute);
+		private static SyntaxNode CreateNewAttribute (
```

Comment:
Is this extra method really necessary? It doesn't look like it's used anywhere else

Reply:
Removed.